lib/puppet/face/config.rb

```diff
@@ -159,6 +159,12 @@ def report_section_and_environment(section_name, environment_name)
         report_section_and_environment(options[:section], Puppet.settings[:environment])
       end
 
+      # Validate setting if it has a hook
+      setting = Puppet.settings.setting(name.to_sym)
+      if setting && setting.has_hook?
+        setting.handle(value)
+      end
```

Comment:
I thought about creating a temporary settings object and then just setting values on it, so that changes to the temporary settings object don't affect the currently running application. But that can't be prevented due to the way that hooks change global state. For example, the hook for the `log_level` setting changes the global logging level which may affect the outcome of `Puppet::Util::Log.sendlevel?(:info)` below. Also the hook for the `path` setting modifies the global `$LOAD_PATH`.

Reply:
Would've been easier if we had a separate parameter for validating settings instead of `:hook`. Right now there's no limit to what a hook can do, from modifying global state, to changing other settings and whatnot.

Another shortcoming is when we set a setting of a predefined type (e.g. boolean), which has its own munge method defined. (I saw you noticed this as well with the EnumSetting type)

```sh-session
> puppet config set strict_hostname_checking aaaa
Warning: Setting strict_hostname_checking to false is deprecated and will be removed in a future release. Please use regular expressions in your node declarations or explicit fact matching for classification (though be warned that fact based classification may be considered insecure).
   (location: /home/gabi/repo/puppet/lib/puppet/defaults.rb:1538:in `block in initialize_default_settings!')
> cat ~/.puppetlabs/etc/puppet/puppet.conf
[main]
strict_hostname_checking = aaaa
> puppet config print strict_hostname_checking
Error: Could not initialize global default settings: Invalid value '"aaaa"' for boolean parameter: strict_hostname_checking
```

In this case the hook is called (which prints the deprecation warning), but [munge](https://github.com/puppetlabs/puppet/blob/e06b87a71b09e7598182c9753f6151bd4155e659/lib/puppet/settings/boolean_setting.rb#L20) is called only when the setting is read. To circumvent this, `munge` can be called from inside the hook, but it's not really a solution...

Reply:
To make matters weirder, I just found out that `munge` is actually only called if the setting happens to have a hook... as it's called by `interpolate` here:
https://github.com/puppetlabs/puppet/blob/e06b87a71b09e7598182c9753f6151bd4155e659/lib/puppet/settings.rb#L625

So if you remove the hook from `strict_hostname_checking` the setting will never be validated?

Reply:
And also weirder, some settings classes have a `validate` method, but it's dead code! https://github.com/puppetlabs/puppet/blob/main/lib/puppet/settings/string_setting.rb#L7
https://github.com/puppetlabs/puppet/blob/main/lib/puppet/settings/file_setting.rb#L177

It seems wrong to wait until the value is interpolated in order to validate and munge it. What if instead we moved the `munge` logic into a `convert` method that was implemented consistently for the different settings types. For example, for boolean, enum, etc types.

```ruby
class Puppet::Settings::EnumSetting ...
  ...
  def convert(value)
    if values.include?(value)
      super(value)
    else
      raise Puppet::Settings::ValidationError, ... unless 
    end
  end
```

Then in the `BaseSettings#convert` method, have it call a `validate` hook if one has been defined, passing the setting object and the value to be set.

```ruby
class Puppet::Settings::BaseSetting
  def convert(value)
    if @validate_hook
     @validate_hook.call(self, value)
  end
end
```

Most defaults would just need to specify a `:type => :enum` and get the appropriate validation, e.g. enum, integer, etc. In cases where a setting has additional requirements, like `certname` must be lowercase, then allow it to implement the validate hook:

```ruby
:validate => proc |setting, value| 
  raise ArgumentError, "Certificate names must be lower case") unless value == value.downcase
```
 
I think that would eliminate most cases where we're using `on_*` hooks for validation. The only cases left would be situations where we really do want hooks to perform side-effects, like changing the log level, changing facter search path, etc. And `puppet config set` would be able to validate settings:

```ruby
Puppet.settings.setting(:certname).validate(value)
```

Thoughts?

Reply:
Sounds good to me!

Reply:
Unfortunately that approach won't work, because puppet stores the uninterpolated value of a setting in its `@config`, for example `@config[:splaylimit] = "$runinterval"`. So whenever `runinterval` changes so does `splaylimit`:

```
irb(main):001:0> require 'puppet'
=> true
irb(main):002:0> Puppet.initialize_settings
=> {}
irb(main):003:0> Puppet[:runinterval]
=> 1800
irb(main):004:0> Puppet[:splaylimit]
=> 1800
irb(main):005:0> Puppet[:runinterval] = 5
=> 5
irb(main):006:0> Puppet[:splaylimit]
=> 5
```

Similarly, you can set the value of a property to an uninterpolated value like:

```
$ bx puppet config set ca_server '$server'
$ bx puppet config print ca_server server
ca_server = puppet
server = puppet
```

In order to validate when a property is written, I think we'd need to keep track of both the interpolated and uninterpolated values, and the dependency graph of settings, so that when  `runinterval` is changed, we know to invalidate and update `splaylimit`. But that would be a significant change I think, so for now, I think we're stuck interpolating and validating settings lazily when they're read.